File: ios/Mobile Client/Mobile Client/entities/Store.swift
//
//  Store.swift
//  Mobile Client
//
//  Created by Eric Ruck on 2/19/18.
//  Copyright © 2018 AuditPro. All rights reserved.
//

import Foundation


/**
 * Provides information about a store that can be audited.
 * @author Eric Ruck
 */
class Store {
	private let _clientId: Int
	private let _chainId: Int
	private let _storeId: Int
	private let _chainName: String
	private let _chainCode: String
	private let _storeName: String
	private let _storeIdentifier: String?
	private let _storeAddress: String?
	private let _storeAddress2: String?
	private let _storeCity: String?
	private let _storeZip: String?
	private let _storeLat: Double?
	private let _storeLon: Double?
	private var _history = [AuditHistory]()

	init(clientId: Int, chainId: Int, storeId: Int, chainName: String, chainCode: String,
		storeName: String, storeIdentifier: String?, storeAddress: String?, storeAddress2: String?,
		storeCity: String?, storeZip: String?, storeLat: Double?, storeLon: Double?, history: String?) {
		self._clientId = clientId
		self._chainId = chainId
		self._storeId = storeId
		self._chainName = chainName
		self._chainCode = chainCode
		self._storeName = storeName
		self._storeIdentifier = storeIdentifier
		self._storeAddress = storeAddress
		self._storeAddress2 = storeAddress2
		self._storeCity = storeCity
		self._storeZip = storeZip
		self._storeLat = storeLat
		self._storeLon = storeLon
		self._history = AuditHistory.fromString(history) ?? [AuditHistory]()
	}

	// Getters
	var clientId: Int { return _clientId }
	var chainId: Int { return _chainId }
	var storeId: Int { return _storeId }
	var chainName: String { return _chainName }
	var chainCode: String { return _chainCode }
	var storeName: String { return _storeName }
	var storeIdentifier: String? { return _storeIdentifier }
	var storeAddress: String? { return _storeAddress }
	var storeAddress2: String? { return _storeAddress2 }
	var storeCity: String? { return _storeCity }
	var storeZip: String? { return _storeZip }
	var storeLat: Double? { return _storeLat }
	var storeLon: Double? { return _storeLon }
	var history: String? { return AuditHistory.toString(_history) }

	/**
	 * Formats the city, state and zip as a single line.
	 */
	var cityStateZip: String {
		var res = _storeCity
		if !(_storeZip?.trimmingCharacters(in: .whitespaces).isEmpty ?? true) {
			res = ((res?.trimmingCharacters(in: .whitespaces).isEmpty ?? true) ? "" : res! + " ") + _storeZip!
		}
		return res ?? ""
	}

	/**
	 * Gets the store description for display.
	 */
	var description: String {
		return (_storeIdentifier == nil)
			? _storeName
			: String(format: "%@ (%@)", _storeName, _storeIdentifier!)
	}

	/**
	 * Indicates if this store is geocoded.
	 */
	var isGeocoded: Bool {
		return (_storeLat != nil) && (_storeLon != nil)
	}

	/**
	 * Gets the number of history entries for this store.
	 * @return Audit history entry count
	 */
	var historyCount: Int {
		return _history.count
	}

	/**
	 * Gets the story history at the indicated index in the list.
	 * @param index Audit history list index
	 * @return Audit history entry or null if index is invalid
	 */
	func getHistoryAt(index: Int) -> AuditHistory? {
		return index >= 0 && index < _history.count ? _history[index] : nil
	}

	/**
	 * Add a history entry from the passed source.
	 * @param source JSON source contains audit history fields
	 * @return Success flag
	 */
	func addAuditHistory(json source: [String: Any?]?) -> Bool {
		// Parse the source
		if let parsed = AuditHistory(source) {
			// Apply valid history
			_history.append(parsed)
			return true
		}

		// Failed to parse history
		return false
	}
}

File: android/mobileclient/app/src/main/java/com/auditpro/mobile_client/entities/Store.java
/*
  AuditPRO Mobile Client Android
  Copyright 2018 AuditPRO All Rights Reserved
 */
package com.auditpro.mobile_client.entities;

import com.auditpro.mobile_client.database.StoreRecord;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;


/**
 * Provides information about a store that can be audited.
 * @author Eric Ruck
 */
public class Store {

	protected Store() {}

	public Store(StoreRecord source) {
		setClientId(source.getClientId());
		setChainId(source.getChainId());
		setChainName(source.getChainName());
		setChainCode(source.getChainCode());
		setStoreId(source.getStoreId());
		setStoreName(source.getStoreName());
		setStoreIdentifier(source.getStoreIdentifier());
		setStoreAddress(source.getStoreAddress());
		setStoreAddress2(source.getStoreAddress2());
		setStoreCity(source.getStoreCity());
		setStoreZip(source.getStoreZip());
		setStoreLat(source.getStoreLat());
		setStoreLon(source.getStoreLon());
		history = AuditHistory.fromString(source.getHistory());
	}

	public int getClientId() {
		return clientId;
	}

	protected void setClientId(int value) {
		clientId = value;
	}

	public int getChainId() {
		return chainId;
	}

	protected void setChainId(int value) {
		chainId = value;
	}

	public String getChainName() {
		return chainName;
	}

	protected void setChainName(String value) {
		chainName = value;
	}

	public String getChainCode() {
		return chainCode;
	}

	protected void setChainCode(String value) {
		chainCode = value;
	}

	public int getStoreId() {
		return storeId;
	}

	protected void setStoreId(int value) {
		storeId = value;
	}

	public String getStoreName() {
		return storeName;
	}

	protected void setStoreName(String value) {
		storeName = value;
	}

	public String getStoreIdentifier() {
		return storeIdentifier;
	}

	protected void setStoreIdentifier(String value) {
		storeIdentifier = value;
	}

	public String getStoreAddress() {
		return storeAddress;
	}

	protected void setStoreAddress(String value) {
		storeAddress = value;
	}

	public String getStoreAddress2() {
		return storeAddress2;
	}

	protected void setStoreAddress2(String value) {
		storeAddress2 = value;
	}

	public String getStoreCity() {
		return storeCity;
	}

	protected void setStoreCity(String value) {
		storeCity = value;
	}

	public String getStoreZip() {
		return storeZip;
	}

	protected void setStoreZip(String value) {
		storeZip = value;
	}

	public Double getStoreLat() {
		return storeLat;
	}

	protected void setStoreLat(Double value) {
		storeLat = value;
	}

	public Double getStoreLon() {
		return storeLon;
	}

	protected void setStoreLon(Double value) {
		storeLon = value;
	}

	public boolean hasLatLong() {
		return (storeLat != null) && (storeLon != null);
	}

	/**
	 * Gets the store description for display.
	 * @return Formatted description to display
	 */
	public String getDescription() {
		String res = (storeName == null) ? chainName : storeName;
		if (storeIdentifier != null) {
			if (res == null) {
				res = storeIdentifier;
			} else {
				res += String.format(" (%s)", storeIdentifier);
			}
		}
		return (res == null) ? String.format(Locale.getDefault(), "%d", storeId) : res;
	}

	@Override public String toString() {
		return getDescription();
	}

	/**
	 * Formats the city, state and zip as a single line.
	 * @return Formatted address
	 */
	public String getCityStateZip() {
		String res = storeCity;
		if ((storeZip != null) && !storeZip.matches("^\\s*$")) {
			res = (res.matches("^\\s*$") ? "" : res + " ") + storeZip;
		}
		return (res == null) ? "" : res;
	}

	/**
	 * Indicates if this store is geocoded.
	 * @return Geocoded flag
	 */
	public boolean isGeocoded() {
		return (getStoreLat() != null) && (getStoreLon() != null);
	}

	/**
	 * Gets the number of history entries for this store.
	 * @return Audit history entry count
	 */
	public int getHistoryCount() {
		return history.size();
	}

	/**
	 * Gets the story history at the indicated index in the list.
	 * @param index Audit history list index
	 * @return Audit history entry or null if index is invalid
	 */
	public AuditHistory getHistory(int index) {
		return ((index >= 0) && (index < history.size())) ? history.get(index) : null;
	}

	/**
	 * Gets an iterable view of all the history entries.
	 * @return Iterable histories
	 */
	public Iterable<AuditHistory> getHistory() {
		return history;
	}

	/**
	 * Add a history entry from the passed source.
	 * @param source JSON source contains audit history fields
	 * @return Success flag
	 */
	protected boolean addAuditHistory(JSONObject source) {
		// Parse the source
		AuditHistory parsed = AuditHistory.fromJSON(source);
		if (parsed == null) {
			// Failed to parse
			return false;
		}

		// Add the history
		history.add(parsed);
		return true;
	}

	private int clientId;
	private int chainId;
	private String chainName;
	private String chainCode;
	private int storeId;
	private String storeName;
	private String storeIdentifier;
	private String storeAddress;
	private String storeAddress2;
	private String storeCity;
	private String storeZip;
	private Double storeLat;
	private Double storeLon;
	private List<AuditHistory> history = new ArrayList<>();
}
